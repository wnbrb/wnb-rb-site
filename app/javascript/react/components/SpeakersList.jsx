import React from 'react';
import PropTypes from 'prop-types';

const SpeakersList = ({ speakers }) => {
    if (speakers?.length > 0) {
        return speakers.map(({ id, name, tagline, image_url }) => (
            <div key={id} className="flex content-center mb-8 text-lg">
                <img className="object-cover w-14 h-14 mr-4 rounded-full" src={image_url} alt="" />
                <div>
                    <p className="font-bold text-gray md:text-lg">{name}</p>
                    <p className="text-sm text-gray md:text-lg">{tagline}</p>
                </div>
            </div>
        ));
    } else {
        return null;
    }
};

SpeakersList.propTypes = {
    speakers: PropTypes.arrayOf(PropTypes.object),
};

export default SpeakersList;
