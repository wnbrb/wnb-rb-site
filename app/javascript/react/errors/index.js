export class UnauthorizedError extends Error {
    constructor(message) {
        super(message);
        this.name = 'UnauthorizedError';
    }
}
