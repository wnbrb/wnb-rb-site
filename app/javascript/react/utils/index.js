const MS_IN_DAY = 24 * 60 * 60 * 1000;

export const postedAtString = (createdAtString) => {
    const today = new Date();
    const postedDate = new Date(createdAtString);

    const diff = (+today - +postedDate) / MS_IN_DAY;

    if (diff <= 1) {
        return 'today';
    }

    if (diff < 7) {
        return `${Math.floor(diff)} days ago`;
    }

    if (diff < 30) {
        const weekDiff = Math.floor(diff / 7);
        const weekOrWeeks = weekDiff === 1 ? 'week' : 'weeks';

        return `${weekDiff} ${weekOrWeeks} ago`;
    }

    const monthDiff = Math.floor(diff / 30);
    const monthOrMonths = monthDiff === 1 ? 'month' : 'months';

    return `${monthDiff} ${monthOrMonths} ago`;
};
