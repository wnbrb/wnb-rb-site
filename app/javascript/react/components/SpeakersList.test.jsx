import React from 'react';
import { render } from '@testing-library/react';
import SpeakersList from './SpeakersList';

describe('SpeakersList Component', () => {
    const mockSpeakers = [
        {
            id: 1,
            name: 'Jane Doe',
            tagline: 'Frontend Developer',
            image_url: 'https://example.com/jane.jpg',
        },
        {
            id: 2,
            name: 'John Smith',
            tagline: 'Backend Engineer',
            image_url: 'https://example.com/john.jpg',
        },
    ];

    const singleSpeaker = [
        {
            id: 1,
            name: 'Alice Johnson',
            tagline: 'Full Stack Developer',
            image_url: 'https://example.com/alice.jpg',
        },
    ];

    test('renders speakers list with multiple speakers', () => {
        const { container } = render(<SpeakersList speakers={mockSpeakers} />);
        expect(container.firstChild).toMatchSnapshot();
    });

    test('renders speakers list with single speaker', () => {
        const { container } = render(<SpeakersList speakers={singleSpeaker} />);
        expect(container.firstChild).toMatchSnapshot();
    });

    test('renders null when speakers array is empty', () => {
        const { container } = render(<SpeakersList speakers={[]} />);
        expect(container.firstChild).toBeNull();
    });

    test('renders null when speakers is undefined', () => {
        const { container } = render(<SpeakersList />);
        expect(container.firstChild).toBeNull();
    });

    test('renders null when speakers is null', () => {
        const { container } = render(<SpeakersList speakers={null} />);
        expect(container.firstChild).toBeNull();
    });

    test('displays speaker names correctly', () => {
        const { getByText } = render(<SpeakersList speakers={mockSpeakers} />);

        expect(getByText('Jane Doe')).toBeInTheDocument();
        expect(getByText('John Smith')).toBeInTheDocument();
    });

    test('displays speaker taglines correctly', () => {
        const { getByText } = render(<SpeakersList speakers={mockSpeakers} />);

        expect(getByText('Frontend Developer')).toBeInTheDocument();
        expect(getByText('Backend Engineer')).toBeInTheDocument();
    });

    test('renders speaker images with correct attributes', () => {
        const { container } = render(<SpeakersList speakers={singleSpeaker} />);

        const image = container.querySelector('img');
        expect(image).toHaveAttribute('src', 'https://example.com/alice.jpg');
        expect(image).toHaveAttribute('alt', '');
        expect(image).toHaveClass('object-cover', 'w-14', 'h-14', 'mr-4', 'rounded-full');
    });

    test('applies correct CSS classes to speaker containers', () => {
        const { container } = render(<SpeakersList speakers={singleSpeaker} />);

        const speakerDiv = container.firstChild;
        expect(speakerDiv).toHaveClass('flex', 'content-center', 'mb-8', 'text-lg');
    });

    test('applies correct CSS classes to speaker names', () => {
        const { getByText } = render(<SpeakersList speakers={singleSpeaker} />);

        const nameElement = getByText('Alice Johnson');
        expect(nameElement).toHaveClass('font-bold', 'text-gray', 'md:text-lg');
    });

    test('applies correct CSS classes to speaker taglines', () => {
        const { getByText } = render(<SpeakersList speakers={singleSpeaker} />);

        const taglineElement = getByText('Full Stack Developer');
        expect(taglineElement).toHaveClass('text-sm', 'text-gray', 'md:text-lg');
    });

    test('uses speaker id as key for React elements', () => {
        const { container } = render(<SpeakersList speakers={mockSpeakers} />);

        // Check that multiple speaker elements are rendered
        const speakerElements = container.querySelectorAll('.flex.content-center');
        expect(speakerElements).toHaveLength(2);
    });
});
