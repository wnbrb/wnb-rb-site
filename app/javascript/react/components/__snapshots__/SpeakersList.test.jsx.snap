// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpeakersList Component renders speakers list with multiple speakers 1`] = `
<div
  class="flex content-center mb-8 text-lg"
>
  <img
    alt=""
    class="object-cover w-14 h-14 mr-4 rounded-full"
    src="https://example.com/jane.jpg"
  />
  <div>
    <p
      class="font-bold text-gray md:text-lg"
    >
      Jane Doe
    </p>
    <p
      class="text-sm text-gray md:text-lg"
    >
      Frontend Developer
    </p>
  </div>
</div>
`;

exports[`SpeakersList Component renders speakers list with single speaker 1`] = `
<div
  class="flex content-center mb-8 text-lg"
>
  <img
    alt=""
    class="object-cover w-14 h-14 mr-4 rounded-full"
    src="https://example.com/alice.jpg"
  />
  <div>
    <p
      class="font-bold text-gray md:text-lg"
    >
      Alice Johnson
    </p>
    <p
      class="text-sm text-gray md:text-lg"
    >
      Full Stack Developer
    </p>
  </div>
</div>
`;
